Added how-to guides (#8624)
Qiskit/qiskit: docs/how_to/compose_quantum_circuits.rst

########################
Compose quantum circuits
########################

> 

Author: I didn't know you could compose with `+` in the past. I can add a warning about that.

> 

Reviewer: FYI. Its part of upgrade notes here https://qiskit.org/documentation/release_notes.html#release-notes-0-23-0-upgrade-notes

>The [QuantumCircuit](https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.html#qiskit.circuit.QuantumCircuit) methods combine and extend have been removed following their deprecation in Qiskit Terra 0.17.0. This was done because these functions were simply less powerful versions of [QuantumCircuit.compose()](https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.compose.html#qiskit.circuit.QuantumCircuit.compose), which should be used instead.
>  The removal of extend also means that the + and += operators are no longer defined for [QuantumCircuit](https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.html#qiskit.circuit.QuantumCircuit). Instead, you can use the & and &= operators respectively, which use [QuantumCircuit.compose()](https://qiskit.org/documentation/stubs/qiskit.circuit.QuantumCircuit.compose.html#qiskit.circuit.QuantumCircuit.compose).

=================

You can join the circuits together using these two methods:

1. Using the :meth:`~.QuantumCircuit.compose` method
2. Using the :meth:`~.QuantumCircuit.append` method

.. warning::

     The ``combine`` and ``extend`` methods have been deprecated in Qiskit Terra 0.17 and removed in 0.23. These methods are replaced by the :meth:`~.QuantumCircuit.compose` method which is more powerful. The removal of ``extend`` also means that the ``+`` and ``+=`` operators are no longer defined for :class:`~.QuantumCircuit`. Instead, you can use the ``&`` and ``&=`` operators respectively, which use :meth:`~.QuantumCircuit.compose`.

For both methods, if the two circuits being combined have different sizes, the method needs to be called in the circuit that is bigger (more qubits and more classical bits). 

Using the :meth:`~.QuantumCircuit.compose` method
-------------------------------------------------

In order to join two circuits with :meth:`~.QuantumCircuit.compose`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 

> By default, :meth:`~.QuantumCircuit.compose` does not modify the original circuit to which it is applied but returns a new joint circuit object. This can be changed by setting the ``inplace`` argument to ``True``.

Reviewer 2: I think it would be good to put this sentence in a note box or a in brackets, as the following code example doesn't actually show how to use the inplace argument

Reviewer 2: alternatively it might be nice to add a short code snippet showing the use of the inplace arg

              ┌─┴─┐┌─┐┌───┐  │   ║ 
    q_3: ─────┤ X ├┤M├┤ H ├──■───╫─
              └───┘└╥┘└───┘      ║ 
    c: 2/═══════════╩════════════╩═
                    1            1 

You can also join the smaller circuit in front of the bigger circuit by setting the ``front`` argument to ``True``.

.. testcode::

    qc5 = qc2.compose(qc1, front=True)
    print(qc5.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

         ┌───┐     ┌───┐             
    q_0: ┤ H ├──■──┤ Y ├───────■─────
         └───┘┌─┴─┐└┬─┬┘┌───┐  │     
    q_1: ─────┤ X ├─┤M├─┤ X ├──┼─────
              └───┘ └╥┘ └───┘  │     
    q_2: ────────────╫─────────┼─────
                     ║       ┌─┴─┐┌─┐
    q_3: ────────────╫───────┤ X ├┤M├
                     ║       └───┘└╥┘
    c: 2/════════════╩═════════════╩═
                     0             1 

Using the :meth:`~.QuantumCircuit.append` method
------------------------------------------------

In order to join two circuits with :meth:`~.QuantumCircuit.append`, you need to specify the circuit you want to add, as well as the qubits and classical bits (if there are any) onto which you want the circuit to be applied.

> Different from :meth:`~.QuantumCircuit.compose`, this method modifies the circuit it is applied to, instead of returning a new circuit.

Reviewer 2: ```suggestion
Unlike :meth:`~.QuantumCircuit.compose`, this method modifies the circuit it is applied to, instead of returning a new circuit.
```

#####################
Join quantum circuits
#####################

> This guide shows how to combine different :class:`~.QuantumCircuit` objects.

Reviewer 3: ```suggestion
This guide shows you how to combine different :class:`~.QuantumCircuit` objects.
```

> This guide shows how to combine different :class:`~.QuantumCircuit` objects.

Reviewer 3: ```suggestion
This guide shows you how to join different :class:`~.QuantumCircuit` objects.
```

> Build the circuits
> ==================

Reviewer 3: ```suggestion
Create the circuits
===================
```

The first step is creating the circuits you want to combine.

.. testcode::

    from qiskit import QuantumCircuit

    qc1 = QuantumCircuit(2,1)
    qc1.h(0)
    qc1.cx(0,1)
    qc1.measure(1,0)

    qc2 = QuantumCircuit(4,2)
    qc2.y(0)
    qc2.x(1)
    qc2.cx(0,3)
    qc2.measure(3,1)

    print(qc1.draw()) 
    print(qc2.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

         ┌───┐        
    q_0: ┤ H ├──■─────
         └───┘┌─┴─┐┌─┐
    q_1: ─────┤ X ├┤M├
              └───┘└╥┘
    c: 1/═══════════╩═
                    0 
         ┌───┐        
    q_0: ┤ Y ├──■─────
         ├───┤  │     
    q_1: ┤ X ├──┼─────
         └───┘  │     
    q_2: ───────┼─────
              ┌─┴─┐┌─┐
    q_3: ─────┤ X ├┤M├
              └───┘└╥┘
    c: 2/═══════════╩═
                    1 

> Combine the circuits
> ====================

Reviewer 3: ```suggestion
Join the circuits
=================
```

Now that you have built the circuits, they can be combined with two different methods:

* :meth:`~.QuantumCircuit.compose`
* :meth:`~.QuantumCircuit.append`

.. warning::

     The usage of the ``+`` operator to combine :class:`~.QuantumCircuit`\ s is no longer supported so if you try to define
     a circuit as ``composed_qc = qc1 + qc2`` you will get an error.

One detail these two methods have in common is that if the circuits have different sizes, they have to be applied to the one that has the most of both qubits and classical bits.

:meth:`~.QuantumCircuit.compose`
------------------------------------------------

> In order to combine two circuits with :meth:`~.QuantumCircuit.compose`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 

Reviewer 3: ```suggestion
In order to join two circuits with :meth:`~.QuantumCircuit.compose`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 
```

> In order to combine two circuits with :meth:`~.QuantumCircuit.compose`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 

Reviewer 3: ```suggestion
In order to join two circuits with :meth:`~.QuantumCircuit.compose`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 
```

By default, :meth:`~.QuantumCircuit.compose` does not change the circuit to which it is applied but returns the composed circuit. This can be changed by setting the ``inplace`` argument to ``True``.

.. testcode::

    qc3 = qc2.compose(qc1)
    print(qc3.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

         ┌───┐     ┌───┐        
    q_0: ┤ Y ├──■──┤ H ├──■─────
         ├───┤  │  └───┘┌─┴─┐┌─┐
    q_1: ┤ X ├──┼───────┤ X ├┤M├
         └───┘  │       └───┘└╥┘
    q_2: ───────┼─────────────╫─
              ┌─┴─┐ ┌─┐       ║ 
    q_3: ─────┤ X ├─┤M├───────╫─
              └───┘ └╥┘       ║ 
    c: 2/════════════╩════════╩═
                     1        0 

> If you want to insert the qubits and bits into specific positions in the bigger circuit, you can use the ``qubits`` and ``bits`` arguments.

Reviewer 3: Can you add a sentence to explain that the smaller circuit is now joint in the following way: q0 => q3 and q1=>q1 (smaller circuit index => bigger circuit index).

Reviewer 3: Addressed in https://github.com/Qiskit/qiskit-terra/pull/8624/commits/802b53dbd8c534821d49284accd74403c8886dee

.. testcode::

    qc4 = qc2.compose(qc1, qubits=[3,1], clbits=[1])
    print(qc4.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

         ┌───┐                     
    q_0: ┤ Y ├──■──────────────────
         ├───┤  │          ┌───┐┌─┐
    q_1: ┤ X ├──┼──────────┤ X ├┤M├
         └───┘  │          └─┬─┘└╥┘
    q_2: ───────┼────────────┼───╫─
              ┌─┴─┐┌─┐┌───┐  │   ║ 
    q_3: ─────┤ X ├┤M├┤ H ├──■───╫─
              └───┘└╥┘└───┘      ║ 
    c: 2/═══════════╩════════════╩═
                    1            1 

> You can also apply the gates from the smaller circuit before those of the bigger one setting the ``front`` argument to ``True``.

Reviewer 3: ```suggestion
You can also join the smaller circuit in front of the bigger circuit by setting the ``front`` argument to ``True``.
```

.. testcode::

    qc5 = qc2.compose(qc1, front=True)
    print(qc5.draw())

.. testoutput::
    :options: +NORMALIZE_WHITESPACE

         ┌───┐     ┌───┐             
    q_0: ┤ H ├──■──┤ Y ├───────■─────
         └───┘┌─┴─┐└┬─┬┘┌───┐  │     
    q_1: ─────┤ X ├─┤M├─┤ X ├──┼─────
              └───┘ └╥┘ └───┘  │     
    q_2: ────────────╫─────────┼─────
                     ║       ┌─┴─┐┌─┐
    q_3: ────────────╫───────┤ X ├┤M├
                     ║       └───┘└╥┘
    c: 2/════════════╩═════════════╩═
                     0             1 

> :meth:`~.QuantumCircuit.append`
> -----------------------------------------------

Reviewer 3: ```suggestion
Using the :meth:`~.QuantumCircuit.append` method
------------------------------------------------
```

> In order to combine two circuits with :meth:`~.QuantumCircuit.append`, you have to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.

Reviewer 3: ```suggestion
In order to join two circuits with :meth:`~.QuantumCircuit.append`, you need to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.
```

> In order to combine two circuits with :meth:`~.QuantumCircuit.append`, you have to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.

Reviewer 3: ```suggestion
In order to join two circuits with :meth:`~.QuantumCircuit.append`, you need to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.
```

> This method changes the circuit to which it is applied instead of returning another one.

Reviewer 3: ```suggestion
Different from :meth:`~.QuantumCircuit.compose`, this method modifies the circuit to which it is applied instead of returning a new circuit.
```

> This method changes the circuit to which it is applied instead of returning another one.

Reviewer 3: ```suggestion
Different from :meth:`~.QuantumCircuit.compose`, this method modifies the circuit to which it is applied instead of returning a new circuit.
```

.. testcode::

    qc2.append(qc1, qargs=[3,1], cargs=[1])

>     qc2.draw(cregbundle=False)

Reviewer 3: What does `cregbundle=False` do? To show individual classical bits instead of showing them in one line? Can you add a comment saying that? Otherwise people will wonder what it does.

Author: Yes, that's exactly what it does. However I might need to check whether this is still necessary. It's possible that it's no longer needed.

.. code-block:: text

         ┌───┐                        
    q_0: ┤ Y ├──■─────────────────────
         ├───┤  │     ┌──────────────┐
    q_1: ┤ X ├──┼─────┤1             ├
         └───┘  │     │              │
    q_2: ───────┼─────┤              ├
              ┌─┴─┐┌─┐│              │
    q_3: ─────┤ X ├┤M├┤0 circuit-101 ├
              └───┘└╥┘│              │
    c_0: ═══════════╬═╡              ╞
                    ║ │              │
    c_1: ═══════════╩═╡0             ╞
                      └──────────────┘

> Unlike :meth:`~.QuantumCircuit.compose`, :meth:`~.QuantumCircuit.append` turns the smaller circuit into a single :class:`~qiskit.circuit.Instruction`, so in order to unroll it you can use :meth:`~.QuantumCircuit.decompose`

Reviewer 3: ```suggestion
Unlike :meth:`~.QuantumCircuit.compose`, :meth:`~.QuantumCircuit.append` turns the smaller circuit into a single :class:`~qiskit.circuit.Instruction`. If you prefer joining the circuits using the individual gates, you can use :meth:`~.QuantumCircuit.decompose` to decompose the joint circuit.
```

------------------------------------------------

In order to combine two circuits with :meth:`~qiskit.circuit.QuantumCircuit.compose()`, you only have to specify the circuit you want to insert. That way the qubits and bits of the smaller circuit will be included into the first qubits and bits of the bigger one in the original order they had. 

By default, :meth:`~qiskit.circuit.QuantumCircuit.compose()` does not change the circuit to which it is applied but returns the composed circuit. This can be changed by setting the ``inplace`` argument to ``True``.

.. jupyter-execute::

    qc3 = qc2.compose(qc1)
    qc3.draw()

If you want to insert the qubits and bits into specific positions in the bigger circuit, you can use the ``qubits`` and ``bits`` arguments.

.. jupyter-execute::

    qc4 = qc2.compose(qc1, qubits=[3,1], clbits=[1])
    qc4.draw()

You can also apply the gates from the smaller circuit before those of the bigger one setting the ``front`` argument to ``True``.

.. jupyter-execute::

    qc5 = qc2.compose(qc1, front=True)
    qc5.draw()

:meth:`~qiskit.circuit.QuantumCircuit.append()`
-----------------------------------------------

> In order to combine two circuits with :meth:`~qiskit.circuit.QuantumCircuit.append()`, you have to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.

Reviewer 4: You can simplify all your references to something like this:

```suggestion
In order to combine two circuits with :meth:`~.QuantumCircuit.append`, you have to specify the circuit you want to add and the qubits and classical bits (if there are any) into which you want the circuit to be inserted.
```

This method changes the circuit to which it is applied instead of returning another one.

.. jupyter-execute::

    qc2.append(qc1, qargs=[3,1], cargs=[1])
    qc2.draw(cregbundle=False)

Unlike :meth:`~qiskit.circuit.QuantumCircuit.compose()`, :meth:`~qiskit.circuit.QuantumCircuit.append()` turns the smaller circuit into a single :class:`~qiskit.circuit.Instruction`, so in order to unroll it you can use :meth:`~qiskit.circuit.QuantumCircuit.decompose()`

.. jupyter-execute::

    qc2.decompose().draw()

> .. jupyter-execute::
> 
>     import qiskit.tools.jupyter
>     %qiskit_version_table
>     %qiskit_copyright
> 
> 
> 

Reviewer 4: You don't really need this, the sphinx build automatically includes the version number and copyright statement in the page footer

```suggestion
```

Author: Should we keep the version table? It can help to keep track of how current the code is.
